### ISSUE TITLE
Error syncing deployment, replica set already exists

### ISSUE THREAD

**Issue by gigaroby** [kubernetes/kubernetes#29735]
Env: Kubernetes 1.3.3 on AWS

I ran into the same problem described into #26673 but now it's happening on every deploy.
I would have reopened the issue, but I was not the author so I could not.

The bug presents itself because I have a build that does `kubectl rollout status deployment/application-unstable` just after a `kubectl apply -f <new deployment manifest>` and it gets stuck forever because the rollout is never reported as finished (it does finish tho).

```
Jul 26 12:42:27 ip-x-x-x-x kube-controller-manager[17798]: E0726 12:42:27.615754   17798 deployment_controller.go:400] Error syncing deployment default/application-unstable: error creating replica set application-unstable: replicasets "application-unstable-900752437" already exists
Jul 26 12:42:27 ip-x-x-x-x kube-controller-manager[17798]: E0726 12:42:27.557964   17798 deployment_controller.go:400] Error syncing deployment default/application-celery-unstable: error creating replica set application-celery-unstable: replicasets "application-celery-unstable-1891916216" already exists
Jul 26 12:42:27 ip-x-x-x-x kube-controller-manager[17798]: E0726 12:42:27.509632   17798 deployment_controller.go:400] Error syncing deployment default/application-celery-worker-unstable: error creating replica set application-celery-worker-unstable: replicasets "application-celery-worker-unstable-3295993310" already exists
Jul 26 12:42:27 ip-x-x-x-x kube-controller-manager[17798]: E0726 12:42:27.456868   17798 deployment_controller.go:400] Error syncing deployment default/application-unstable: error creating replica set application-unstable: replicasets "application-unstable-900752437" already exists
Jul 26 12:42:27 ip-x-x-x-x kube-controller-manager[17798]: E0726 12:42:27.409374   17798 deployment_controller.go:400] Error syncing deployment default/application-celery-unstable: error creating replica set application-celery-unstable: replicasets "application-celery-unstable-1891916216" already exists
Jul 26 12:42:27 ip-x-x-x-x kube-controller-manager[17798]: E0726 12:42:27.357072   17798 deployment_controller.go:400] Error syncing deployment default/application-celery-worker-unstable: error creating replica set application-celery-worker-unstable: replicasets "application-celery-worker-unstable-3295993310" already exists
Jul 26 12:42:27 ip-x-x-x-x kube-controller-manager[17798]: E0726 12:42:27.306676   17798 deployment_controller.go:400] Error syncing deployment default/application-unstable: error creating replica set application-unstable: replicasets "application-unstable-900752437" already exists
Jul 26 12:42:27 ip-x-x-x-x kube-controller-manager[17798]: E0726 12:42:27.257030   17798 deployment_controller.go:400] Error syncing deployment default/application-celery-unstable: error creating replica set application-celery-unstable: replicasets "application-celery-unstable-1891916216" already exists
Jul 26 12:42:27 ip-x-x-x-x kube-controller-manager[17798]: E0726 12:42:27.206504   17798 deployment_controller.go:400] Error syncing deployment default/application-celery-worker-unstable: error creating replica set application-celery-worker-unstable: replicasets "application-celery-worker-unstable-3295993310" already exists
Jul 26 12:42:27 ip-x-x-x-x kube-controller-manager[17798]: E0726 12:42:27.157741   17798 deployment_controller.go:400] Error syncing deployment default/application-unstable: error creating replica set application-unstable: replicasets "application-unstable-900752437" already exists
Jul 26 12:42:27 ip-x-x-x-x kube-controller-manager[17798]: E0726 12:42:27.112898   17798 deployment_controller.go:400] Error syncing deployment default/application-celery-unstable: error creating replica set application-celery-unstable: replicasets "application-celery-unstable-1891916216" already exists
Jul 26 12:42:27 ip-x-x-x-x kube-controller-manager[17798]: E0726 12:42:27.059652   17798 deployment_controller.go:400] Error syncing deployment default/application-celery-worker-unstable: error creating replica set application-celery-worker-unstable: replicasets "application-celery-worker-unstable-3295993310" already exists
Jul 26 12:42:27 ip-x-x-x-x kube-controller-manager[17798]: E0726 12:42:27.007476   17798 deployment_controller.go:400] Error syncing deployment default/application-unstable: error creating replica set application-unstable: replicasets "application-unstable-900752437" already exists
Jul 26 12:42:26 ip-x-x-x-x kube-controller-manager[17798]: E0726 12:42:26.956218   17798 deployment_controller.go:400] Error syncing deployment default/application-celery-unstable: error creating replica set application-celery-unstable: replicasets "application-celery-unstable-1891916216" already exists
Jul 26 12:42:26 ip-x-x-x-x kube-controller-manager[17798]: E0726 12:42:26.910275   17798 deployment_controller.go:400] Error syncing deployment default/application-celery-worker-unstable: error creating replica set application-celery-worker-unstable: replicasets "application-celery-worker-unstable-3295993310" already exists
Jul 26 12:42:26 ip-x-x-x-x kube-controller-manager[17798]: E0726 12:42:26.857240   17798 deployment_controller.go:400] Error syncing deployment default/application-unstable: error creating replica set application-unstable: replicasets "application-unstable-900752437" already exists
Jul 26 12:42:26 ip-x-x-x-x kube-controller-manager[17798]: E0726 12:42:26.806429   17798 deployment_controller.go:400] Error syncing deployment default/application-celery-unstable: error creating replica set application-celery-unstable: replicasets "application-celery-unstable-1891916216" already exists
Jul 26 12:42:26 ip-x-x-x-x kube-controller-manager[17798]: E0726 12:42:26.756885   17798 deployment_controller.go:400] Error syncing deployment default/application-celery-worker-unstable: error creating replica set application-celery-worker-unstable: replicasets "application-celery-worker-unstable-3295993310" already exists
Jul 26 12:42:26 ip-x-x-x-x kube-controller-manager[17798]: E0726 12:42:26.706639   17798 deployment_controller.go:400] Error syncing deployment default/application-unstable: error creating replica set application-unstable: replicasets "application-unstable-900752437" already exists
Jul 26 12:42:26 ip-x-x-x-x kube-controller-manager[17798]: E0726 12:42:26.656417   17798 deployment_controller.go:400] Error syncing deployment default/application-celery-unstable: error creating replica set application-celery-unstable: replicasets "application-celery-unstable-1891916216" already exists
Jul 26 12:42:26 ip-x-x-x-x kube-controller-manager[17798]: E0726 12:42:26.607126   17798 deployment_controller.go:400] Error syncing deployment default/application-celery-worker-unstable: error creating replica set application-celery-worker-unstable: replicasets "application-celery-worker-unstable-3295993310" already exists
Jul 26 12:42:26 ip-x-x-x-x kube-controller-manager[17798]: E0726 12:42:26.557562   17798 deployment_controller.go:400] Error syncing deployment default/application-unstable: error creating replica set application-unstable: replicasets "application-unstable-900752437" already exists
Jul 26 12:42:26 ip-x-x-x-x kube-controller-manager[17798]: E0726 12:42:26.506025   17798 deployment_controller.go:400] Error syncing deployment default/application-celery-unstable: error creating replica set application-celery-unstable: replicasets "application-celery-unstable-1891916216" already exists
Jul 26 12:42:26 ip-x-x-x-x kube-controller-manager[17798]: E0726 12:42:26.456588   17798 deployment_controller.go:400] Error syncing deployment default/application-celery-worker-unstable: error creating replica set application-celery-worker-unstable: replicasets "application-celery-worker-unstable-3295993310" already exists
Jul 26 12:42:26 ip-x-x-x-x kube-controller-manager[17798]: E0726 12:42:26.422213   17798 deployment_controller.go:400] Error syncing deployment default/application-unstable: error creating replica set application-unstable: replicasets "application-unstable-900752437" already exists
Jul 26 12:42:26 ip-x-x-x-x kube-controller-manager[17798]: E0726 12:42:26.356907   17798 deployment_controller.go:400] Error syncing deployment default/application-celery-unstable: error creating replica set application-celery-unstable: replicasets "application-celery-unstable-1891916216" already exists
Jul 26 12:42:26 ip-x-x-x-x kube-controller-manager[17798]: E0726 12:42:26.313353   17798 deployment_controller.go:400] Error syncing deployment default/application-celery-worker-unstable: error creating replica set application-celery-worker-unstable: replicasets "application-celery-worker-unstable-3295993310" already exists
Jul 26 12:42:26 ip-x-x-x-x kube-controller-manager[17798]: E0726 12:42:26.257953   17798 deployment_controller.go:400] Error syncing deployment default/application-unstable: error creating replica set application-unstable: replicasets "application-unstable-900752437" already exists
Jul 26 12:42:26 ip-x-x-x-x kube-controller-manager[17798]: E0726 12:42:26.209894   17798 deployment_controller.go:400] Error syncing deployment default/application-celery-unstable: error creating replica set application-celery-unstable: replicasets "application-celery-unstable-1891916216" already exists
Jul 26 12:42:26 ip-x-x-x-x kube-controller-manager[17798]: E0726 12:42:26.160954   17798 deployment_controller.go:400] Error syncing deployment default/application-celery-worker-unstable: error creating replica set application-celery-worker-unstable: replicasets "application-celery-worker-unstable-3295993310" already exists
Jul 26 12:42:26 ip-x-x-x-x kube-controller-manager[17798]: E0726 12:42:26.107205   17798 deployment_controller.go:400] Error syncing deployment default/application-unstable: error creating replica set application-unstable: replicasets "application-unstable-900752437" already exists
Jul 26 12:42:26 ip-x-x-x-x kube-controller-manager[17798]: E0726 12:42:26.056387   17798 deployment_controller.go:400] Error syncing deployment default/application-celery-unstable: error creating replica set application-celery-unstable: replicasets "application-celery-unstable-1891916216" already exists
Jul 26 12:42:26 ip-x-x-x-x kube-controller-manager[17798]: E0726 12:42:26.006022   17798 deployment_controller.go:400] Error syncing deployment default/application-celery-worker-unstable: error creating replica set application-celery-worker-unstable: replicasets "application-celery-worker-unstable-3295993310" already exists
Jul 26 12:42:25 ip-x-x-x-x kube-controller-manager[17798]: E0726 12:42:25.956830   17798 deployment_controller.go:400] Error syncing deployment default/application-unstable: error creating replica set application-unstable: replicasets "application-unstable-900752437" already exists
Jul 26 12:42:25 ip-x-x-x-x kube-controller-manager[17798]: E0726 12:42:25.906450   17798 deployment_controller.go:400] Error syncing deployment default/application-celery-unstable: error creating replica set application-celery-unstable: replicasets "application-celery-unstable-1891916216" already exists
```

**Comment by AdoHe** [comment]
@gigaroby could you provide more detail info? like what's your resource file and how to reproduce it.

**Comment by gigaroby** [comment]
@AdoHe yes, of course:
We have an internal tool that downloads some manifests from S3 in a temporary folder and then calls `kubectl apply -f temp_dir`. It's harder to narrow it down because that's all there is to it.
It caused the issue above for 3/4 builds in a row but in the last build it worked.
The three deployment files are fairly standard deployment templates, this is an example of one of them:
```
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: application-unstable
  labels:
    project: application-unstable
spec:
  replicas: 2
  template:
    metadata:
      labels:
        app: application-revenge
        environment: unstable
        tier: frontend
    spec:
      containers:
        - name: uwsgi
          image: my-index.domain.com/application-unstable:{{ version }}
          env: []
          command:
            - "uwsgi"
          args:
            - "--ini"
            - "/etc/uwsgi.ini"
            - "--processes"
            - "1"
            - "--threads"
            - "3"

        - name: openresty
          image: spaziodati/openresty:latest
          ports:
            - containerPort: 8080
              name: http
              protocol: "TCP"
          command:
            - nginx
          args:
            - -c
            - /etc/conf/nginx.conf
          volumeMounts:
            - name: config-volume
              mountPath: /etc/conf
          livenessProbe:
            httpGet:
              path: /status
              port: 8080
            initialDelaySeconds: 10
            timeoutSeconds: 3

      volumes:
        - name: config-volume
          configMap:
            name: application-openresty-conf
```
The `{{ version }}` is filled by jenkins at build time before the manifest is uploaded on s3.

I am not sure about how to narrow this down further, any suggestions?

**Comment by kargakis** [comment]
@kubernetes/deployment can somebody triage this?

**Comment by koombea-rops** [comment]
I'm seeing this now, on kube 1.3.6 on AWS. The cluster ran fine for two weeks, and then problem started overnight, and shows no indication of fixing itself (12+ hours later).

Our deploy process is a series of `kubectl apply -f`, run from a bash script during CI

**Comment by arohner** [comment]
We just started seeing this as well. kube 1.3.6 on AWS. 

The cluster was running fine for two weeks, then the problem started on the last CI build yesterday, and shows no signs of fixing itself. Our deploy process is a series of `kubectl apply -f` commands called from a bash script during CI.

**Comment by kargakis** [comment]
https://github.com/kubernetes/kubernetes/pull/28684 should fix the hotloop by using a rate limiter in the deployment controller. Seems like it was included in 1.3.7. Also the "already exists" shouldn't be treated as an error but instead we should return the replica set from the cache at this point (since it already exists). I was handling this error in a recent PR of mine, will update this issue back with the #

**Comment by jsravn** [comment]
@kargakis Are you sure the fix is in 1.3.7? I just hit this problem in 1.3.7, and it kills the master node as it pegs a cpu at 100% in this error loop.

**Comment by kargakis** [comment]
Hm, git fooled me. The fix is included in 1.4.

**Comment by jsravn** [comment]
From the code, it looks like deployment api generates an adler-32 checksum of the pod template, then uses that to create a new replicaset `deploymentname-<hash>`. If that hash already exists, it barfs with the controller-manager loop.  So it looks to me, from a quick read, as simply a hash collision handled poorly. This seems to mirror what I'm seeing, since it only happens on deployments with a history >100 deploys. I'm guessing there isn't enough data change in the pod spec to ensure a decent amount of uniqueness with adler-32 - in my case just the version changes. It should probably use a better hash function (that doesn't duplicate on such a small number of deployments), and/or handle the collision more gracefully.

**Comment by kargakis** [comment]
@jsravn I don't think you would get a hash collision even with >100 deploys (not sure where Adler starts to fail though). Even the slightest change in the pod template changes the hash. Can you provide the pod template of the existing replica set that blocks the deployment controller and also the latest state of your deployment? 

I think this issue is simply us not handling AlreadyExists errors when we try to create new replica sets (the replica set cache may get stale).

**Comment by jsravn** [comment]
@kargakis sure... I admit I may be wrong - that was my just my super quick analysis. :) I need to try reproducing to see if it's actually the case - which I'll try now. Here's the replicaset that it complains already exists:

```yaml
apiVersion: extensions/v1beta1
kind: ReplicaSet
metadata:
  annotations:
    deployment.kubernetes.io/revision: "25"
  creationTimestamp: 2016-09-22T16:00:52Z
  generation: 3
  labels:
    app: cats
    pod-template-hash: "218308181"
  name: cats-218308181
  namespace: cats-stubbed-functional
  resourceVersion: "3958038"
  selfLink: /apis/extensions/v1beta1/namespaces/cats-stubbed-functional/replicasets/cats-218308181
  uid: bcd1bacb-80dd-11e6-bfcc-0a8cd5340591
spec:
  replicas: 0
  selector:
    matchLabels:
      app: cats
      pod-template-hash: "218308181"
  template:
    metadata:
      creationTimestamp: null
      labels:
        app: cats
        pod-template-hash: "218308181"
    spec:
      containers:
      - env:
        - name: DEPLOYMENT_ENVIRONMENT
          value: cats-stubbed-functional
        - name: APP_NAME
          value: cats
        image: registry.tools.cosmic.sky/test/cats:v0.535.0
        imagePullPolicy: IfNotPresent
        livenessProbe:
          failureThreshold: 3
          httpGet:
            path: /private/status
            port: 9077
            scheme: HTTP
          initialDelaySeconds: 30
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 1
        name: cats
        ports:
        - containerPort: 9077
          name: http
          protocol: TCP
        readinessProbe:
          failureThreshold: 3
          httpGet:
            path: /private/status
            port: 9077
            scheme: HTTP
          initialDelaySeconds: 1
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 1
        resources:
          limits:
            cpu: "1"
            memory: 1Gi
          requests:
            cpu: "1"
            memory: 1Gi
        terminationMessagePath: /dev/termination-log
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      securityContext: {}
      terminationGracePeriodSeconds: 30
status:
  observedGeneration: 3
  replicas: 0
```

And the current failing deployment (although, I believe it started failing at generation 204, version 0.612.0).

```yaml
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  annotations:
    deployment.kubernetes.io/revision: "100"
    kubectl.kubernetes.io/last-applied-configuration: '{"kind":"Deployment","apiVersion":"extensions/v1beta1","metadata":{"name":"cats","creationTimestamp":null,"labels":{"app":"cats"}},"spec":{"replicas":1,"template":{"metadata":{"creationTimestamp":null,"labels":{"app":"cats"}},"spec":{"containers":[{"name":"cats","image":"registry.tools.cosmic.sky/test/cats:v0.617.0","ports":[{"name":"http","containerPort":9077,"protocol":"TCP"}],"env":[{"name":"DEPLOYMENT_ENVIRONMENT","value":"cats-stubbed-functional"},{"name":"APP_NAME","value":"cats"}],"resources":{"limits":{"cpu":"1","memory":"1Gi"},"requests":{"cpu":"1","memory":"1Gi"}},"livenessProbe":{"httpGet":{"path":"/private/status","port":9077},"initialDelaySeconds":30,"timeoutSeconds":1},"readinessProbe":{"httpGet":{"path":"/private/status","port":9077},"initialDelaySeconds":1,"timeoutSeconds":1}}]}},"strategy":{"type":"RollingUpdate","rollingUpdate":{"maxUnavailable":0,"maxSurge":1}}},"status":{}}'
  creationTimestamp: 2016-09-11T09:39:41Z
  generation: 204
  labels:
    app: cats
  name: cats
  namespace: cats-stubbed-functional
  resourceVersion: "12695928"
  selfLink: /apis/extensions/v1beta1/namespaces/cats-stubbed-functional/deployments/cats
  uid: a9ceb758-7803-11e6-9e34-0a8cd5340591
spec:
  replicas: 1
  rollbackTo: {}
  selector:
    matchLabels:
      app: cats
  strategy:
    rollingUpdate:
      maxSurge: 1
      maxUnavailable: 0
    type: RollingUpdate
  template:
    metadata:
      creationTimestamp: null
      labels:
        app: cats
    spec:
      containers:
      - env:
        - name: DEPLOYMENT_ENVIRONMENT
          value: cats-stubbed-functional
        - name: APP_NAME
          value: cats
        image: registry.tools.cosmic.sky/test/cats:v0.617.0
        imagePullPolicy: IfNotPresent
        livenessProbe:
          failureThreshold: 3
          httpGet:
            path: /private/status
            port: 9077
            scheme: HTTP
          initialDelaySeconds: 30
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 1
        name: cats
        ports:
        - containerPort: 9077
          name: http
          protocol: TCP
        readinessProbe:
          failureThreshold: 3
          httpGet:
            path: /private/status
            port: 9077
            scheme: HTTP
          initialDelaySeconds: 1
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 1
        resources:
          limits:
            cpu: "1"
            memory: 1Gi
          requests:
            cpu: "1"
            memory: 1Gi
        terminationMessagePath: /dev/termination-log
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      securityContext: {}
      terminationGracePeriodSeconds: 30
status:
  availableReplicas: 1
  observedGeneration: 203
  replicas: 1
  updatedReplicas: 1
```

**Comment by jsravn** [comment]
Hash collisions are definitely a problem. I wrote up a test and can get a collision within a couple hundred versions (only version changes). Here's pod_test.go:

```
package pod

import (
	"testing"
	"strings"
	"strconv"
	"encoding/json"
	"fmt"
	"k8s.io/kubernetes/pkg/api"
)

var podSpec string = `
{
	"metadata": {
		"creationTimestamp": null,
		"labels": {
			"app": "cats"
		}
	},
	"spec": {
		"containers": [
			{
				"name": "cats",
				"image": "registry/test/cats:v0.@@VERSION@@.0",
				"ports": [
					{
						"name": "http",
						"containerPort": 9077,
						"protocol": "TCP"
					}
				],
				"env": [
					{
						"name": "DEPLOYMENT_ENVIRONMENT",
						"value": "cats-stubbed-functional"
					},
					{
						"name": "APP_NAME",
						"value": "cats"
					}
				],
				"resources": {
					"limits": {
						"cpu": "1",
						"memory": "1Gi"
					},
					"requests": {
						"cpu": "1",
						"memory": "1Gi"
					}
				},
				"livenessProbe": {
					"httpGet": {
						"path": "/private/status",
						"port": 9077,
						"scheme": "HTTP"
					},
					"initialDelaySeconds": 30,
					"timeoutSeconds": 1,
					"periodSeconds": 10,
					"successThreshold": 1,
					"failureThreshold": 3
				},
				"readinessProbe": {
					"httpGet": {
						"path": "/private/status",
						"port": 9077,
						"scheme": "HTTP"
					},
					"initialDelaySeconds": 1,
					"timeoutSeconds": 1,
					"periodSeconds": 10,
					"successThreshold": 1,
					"failureThreshold": 3
				},
				"terminationMessagePath": "/dev/termination-log",
				"imagePullPolicy": "IfNotPresent"
			}
		],
		"restartPolicy": "Always",
		"terminationGracePeriodSeconds": 30,
		"dnsPolicy": "ClusterFirst",
		"securityContext": {}
	}
}
`

func TestPodTemplateSpecHash(t *testing.T) {
	seenHashes := make(map[uint32]int)

	for i := 0; i < 1000; i++ {
		specJson := strings.Replace(podSpec, "@@VERSION@@", strconv.Itoa(i), 1)
		spec := api.PodTemplateSpec{}
		json.Unmarshal([]byte(specJson), &spec)
		hash := GetPodTemplateSpecHash(spec)
		fmt.Printf("version: %d hash: %d\n", i, hash)
		if v, ok := seenHashes[hash]; ok {
			t.Fatalf("Hash collision, old: %d new: %d", v, i)
		}
		seenHashes[hash] = i
	}
}
```

I'm guessing we can't improve the hash function at this point, to maintain backwards compatibility? We should at least handle the collision somehow - what do you suggest?

**Comment by jsravn** [comment]
Further, in the example I gave above, the hash output only has a cycle length of 120, after which all newer versions collide with previous hashed values.

```
<snip>
	pod_test.go:98: Hash collision, old: 120 new: 201
	pod_test.go:98: Hash collision, old: 121 new: 202
	pod_test.go:98: Hash collision, old: 122 new: 203
	pod_test.go:98: Hash collision, old: 123 new: 204
	pod_test.go:98: Hash collision, old: 124 new: 205
	pod_test.go:98: Hash collision, old: 125 new: 206
	pod_test.go:98: Hash collision, old: 126 new: 207
	pod_test.go:98: Hash collision, old: 127 new: 208
	pod_test.go:98: Hash collision, old: 128 new: 209
	pod_test.go:98: Hash collision, old: 130 new: 211
	pod_test.go:98: Hash collision, old: 131 new: 212
	pod_test.go:98: Hash collision, old: 132 new: 213
	pod_test.go:98: Hash collision, old: 133 new: 214
	pod_test.go:98: Hash collision, old: 134 new: 215
	pod_test.go:98: Hash collision, old: 135 new: 216
	pod_test.go:98: Hash collision, old: 136 new: 217
	pod_test.go:98: Hash collision, old: 137 new: 218
	pod_test.go:98: Hash collision, old: 138 new: 219
	pod_test.go:98: Hash collision, old: 140 new: 221
	pod_test.go:98: Hash collision, old: 141 new: 222
	pod_test.go:98: Hash collision, old: 142 new: 223
	pod_test.go:98: Hash collision, old: 143 new: 224
	pod_test.go:98: Hash collision, old: 144 new: 225
	pod_test.go:98: Hash collision, old: 145 new: 226
	pod_test.go:98: Hash collision, old: 146 new: 227
	pod_test.go:98: Hash collision, old: 147 new: 228
	pod_test.go:98: Hash collision, old: 148 new: 229
	pod_test.go:98: Hash collision, old: 150 new: 231
	pod_test.go:98: Hash collision, old: 151 new: 232
	pod_test.go:98: Hash collision, old: 152 new: 233
	pod_test.go:98: Hash collision, old: 153 new: 234
	pod_test.go:98: Hash collision, old: 154 new: 235
	pod_test.go:98: Hash collision, old: 155 new: 236
	pod_test.go:98: Hash collision, old: 156 new: 237
	pod_test.go:98: Hash collision, old: 157 new: 238
	pod_test.go:98: Hash collision, old: 158 new: 239
	pod_test.go:98: Hash collision, old: 160 new: 241
	pod_test.go:98: Hash collision, old: 161 new: 242
	pod_test.go:98: Hash collision, old: 162 new: 243
	pod_test.go:98: Hash collision, old: 163 new: 244
	pod_test.go:98: Hash collision, old: 164 new: 245
	pod_test.go:98: Hash collision, old: 165 new: 246
	pod_test.go:98: Hash collision, old: 166 new: 247
	pod_test.go:98: Hash collision, old: 167 new: 248
	pod_test.go:98: Hash collision, old: 168 new: 249
	pod_test.go:98: Hash collision, old: 170 new: 251
	pod_test.go:98: Hash collision, old: 171 new: 252
	pod_test.go:98: Hash collision, old: 172 new: 253
	pod_test.go:98: Hash collision, old: 173 new: 254
	pod_test.go:98: Hash collision, old: 174 new: 255
	pod_test.go:98: Hash collision, old: 175 new: 256
<snip>
```

**Comment by jsravn** [comment]
I suppose the immediate workaround is to limit deployment revision history to small values (http://kubernetes.io/docs/user-guide/deployments/#revision-history-limit).

**Comment by kargakis** [comment]
@jsravn thanks for the test! I run it as well and I see the same results. @smarterclayton @bgrant0607 can you have a look? An immediate workaround is to use revisionHistoryLimit, yes. In reality most users won't care about the history of 200 replica sets but it is concerning that Adler breaks that fast.

**Comment by bgrant0607** [comment]
cc @janetkuo @pwittrock

**Comment by smarterclayton** [comment]
Adler was only used because we were originally overly concerned about the hash performance.  We need to be using a reasonably (but not excessively) fast hash where collision chance is >> 1 in 1e5, and preferably > 1e10 or more.

**Comment by smarterclayton** [comment]
fnv would be an easy change

**Comment by pwittrock** [comment]
Tangentially... having ~200 RS resources hanging around per-deployment seems like a lot and I imagine could result in triggering unknown-unknowns.  IMHO the revision history limit should be bounded by default before moving out of beta.  I can image things like automated hourly builds creating lots of history.

**Comment by smarterclayton** [comment]
Generally bounds would be enforced by limit ranges, which already bound
number of replica sets

On Oct 18, 2016, at 11:31 PM, Phillip Wittrock <notifications@github.com>
wrote:

Tangentially... having ~200 RS resources hanging around per-deployment
seems like a lot and I imagine could result in triggering unknown-unknowns.
IMHO the revision history limit should be bounded by default before moving
out of beta. I can image things like automated hourly builds creating lots
of history.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/29735#issuecomment-254701544>,
or mute the thread

**Comment by pwittrock** [comment]
Thanks

**Comment by bgrant0607** [comment]
cc @lukaszo

**Comment by bgrant0607** [comment]
@pwittrock:

@kargakis sent a PR to change the defaults (#34339), but we'd need a new apiversion to do that.

**Comment by lchi** [comment]
We just ran into this issue with a much smaller history (only 10 RS resources). We're using the revisionHistoryLimit workaround but it seems like the real fix would be using a different hashing algorithm. Any thoughts on when a fix may come out?

**Comment by kargakis** [comment]
@lchi hash collision is not the only case where you can stumble upon this issue. Are you sure this was a hash coliision? Did you check that the deployment pod template is different from the rs pod template that you get in the error?

**Comment by lchi** [comment]
@kargakis yes, here are two templates that we are seeing hash to the same value:

```
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: exworker
spec:
  replicas: 2
  template:
    metadata:
      labels:
        run: exworker
        language: elixir
    spec:
      containers:
      - name: exworker
        image: teacherspayteachers/exworker:release-2.1.2
        env:
          - name: ENVIRONMENT
            value: "prod"
          - name: CORE_API_KEY
            valueFrom:
              secretKeyRef:
                name: core-api-secret-key
                key: secret_key
      imagePullSecrets:
        - name: "tptreadonly-docker"
```

vs

```
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: exworker
spec:
  replicas: 2
  template:
    metadata:
      labels:
        run: exworker
        language: elixir
    spec:
      containers:
      - name: exworker
        image: teacherspayteachers/exworker:release-1.3.1
        env:
          - name: ENVIRONMENT
            value: "prod"
          - name: CORE_API_KEY
            valueFrom:
              secretKeyRef:
                name: core-api-secret-key
                key: secret_key
      imagePullSecrets:
        - name: "tptreadonly-docker"
```

In this case there's only a small change for the version bump, and the template is rather short.

**Comment by smarterclayton** [comment]
Can we even change this in a backwards compatible manner on the cluster
safely?

On Nov 2, 2016, at 5:52 PM, Lucas Chi <notifications@github.com> wrote:

@kargakis <https://github.com/kargakis> yes, here are two templates that we
are seeing hash to the same value:

apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: exworker
spec:
  replicas: 2
  template:
    metadata:
      labels:
        run: exworker
        language: elixir
    spec:
      containers:
      - name: exworker
        image: teacherspayteachers/exworker:release-2.1.2
        env:
          - name: ENVIRONMENT
            value: "prod"
          - name: CORE_API_KEY
            valueFrom:
              secretKeyRef:
                name: core-api-secret-key
                key: secret_key
      imagePullSecrets:
        - name: "tptreadonly-docker"

vs

apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: exworker
spec:
  replicas: 2
  template:
    metadata:
      labels:
        run: exworker
        language: elixir
    spec:
      containers:
      - name: exworker
        image: teacherspayteachers/exworker:release-1.3.1
        env:
          - name: ENVIRONMENT
            value: "prod"
          - name: CORE_API_KEY
            valueFrom:
              secretKeyRef:
                name: core-api-secret-key
                key: secret_key
      imagePullSecrets:
        - name: "tptreadonly-docker"

In this case there's only a small change for the version bump, and the
template is rather short.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/29735#issuecomment-258010535>,
or mute the thread

**Comment by kargakis** [comment]
> Can we even change this in a backwards compatible manner on the cluster safely?

New hash algo? Unlikely.

**Comment by smarterclayton** [comment]
So we don't want to blow out an existing cluster immediately on upgrade, so
we need a way to stage this so an admin can slowly roll this out under
their control.

One possible algo:

1. Add a flag to the controller manager config that decides hash behavior
"legacy", "migrate-to-X", "X"
2. When the deployment controller calculates hashes, calculate old and new
hash
3. If there are no existing replicas for a deployment, use the new hash
4. If the "old" hash matches and a new hash is available, put the
deployment in a queue
5. Pull N at a time from the queue and have them use the new hash
6. Every time a deployment is successfully moved to a new hash, draw a
replacement from the queue
7. After all deployments are migrated, admin can change the hash behavior

We could potentially get away without config, but I think admins need to
opt in to this change in 1.N and be forced to do it in 1.(N+1)

**Comment by kargakis** [comment]
> 3. If there are no existing replicas for a deployment, use the new hash

What about old replica sets? What happens if a new hash collides with an old hash (latest template hashed == old replica set computed with old hashing algo)?

**Comment by smarterclayton** [comment]
Old replica sets is covered by 4.

If you have a hash collision its game over anyway, so not worried about
that.

**Comment by smarterclayton** [comment]
Bumping to P1 because this makes deployments unpredictable - we need
to assess the change required for 1.5 next week and whether we can
come up with a backwards compatible, safe, and non-disruptive change.

**Comment by kargakis** [comment]
This is quite a big change for 1.5 and since the problem was always there I would move a fix for this in 1.6.

**Comment by kargakis** [comment]
Tested [fnv](https://golang.org/pkg/hash/fnv/#New32a) with the unit test provided by @jsravn above and it didn't break for 10^3 up to 10^5 (10^6 was timing out the test and I didn't bother to test more).

**Comment by kargakis** [comment]
Results from benchmarking
```
BenchmarkAdler-4   	   30000	     56445 ns/op
BenchmarkFnv-4     	   30000	     58152 ns/op
```
```
BenchmarkAdler-4   	  500000	     54725 ns/op
BenchmarkFnv-4     	  500000	     59724 ns/op
```
Fnv is a bit slower but much more stable.

**Comment by kargakis** [comment]
@smarterclayton so my understanding is that the migration of old replica sets will happen during the normal operation of the cluster, is that correct? We can migrate all replica sets with zero replicas in the background, and migrate running sets once a new rollout happens. We may not even need a separate queue. I am still not sure if we can notify admins somehow when all existing deployments have migrated.

**Comment by smarterclayton** [comment]
We could expose a metric that is "how many old replication controllers we still have"

**Comment by kargakis** [comment]
For 1.6, we moved Deployments into the apps api group[1] so we can change the defaults (revisionHistoryLimit is now set to 3 by default for newly created deployments) and we also made the cleanup policy run independent of a rollout (so a rollout that gets stuck won't block deletiion of older replica sets)[2]. I've also opened an update to the Deployment proposal for moving away from hashing: https://github.com/kubernetes/community/pull/384

Moving the milestone to 1.7

[1] https://github.com/kubernetes/kubernetes/pull/39683
[2] https://github.com/kubernetes/kubernetes/pull/40081

**Comment by thockin** [comment]
Changing revision history doesn't seem to alleviate hash collisions, right?  Looking at #43449 - are we using a poor hash for this?

**Comment by kargakis** [comment]
> Changing revision history doesn't seem to alleviate hash collisions, right?

Actually, it does. If all you have is 3 old replica sets, it's less likely that the controller will break you because of a hash collision. 200 seems to be the limit with adler.

> Looking at #43449 - are we using a poor hash for this?

Yes. There is https://github.com/kubernetes/kubernetes/pull/38714 that changes us to fnv but there is also https://github.com/kubernetes/community/pull/384 that moves us away from hashing. I am going to open an alternative proposal to 384 that covers the transition to fnv and we need to decide with what do we want to proceed.

**Comment by kargakis** [comment]
Opened https://github.com/kubernetes/community/pull/477 as an alternative to kubernetes/community#384

**Comment by brugz** [comment]
We saw this in a related configuration. Marking --record=true on the deployment [1] and setting the revision history [2] configuration seemed to alleviate the stress. Is there a status on the hash collision bug?

[1] https://kubernetes.io/docs/concepts/workloads/controllers/deployment/ 
[2] https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#revision-history-limit 

Thanks,
Brugz

**Comment by janetkuo** [comment]
@brugz the fix will be included in 1.7 release
